File: haskell/glue/src/Glue/Lib/Math/Trigonometric/Cos.hs
module Glue.Lib.Math.Trigonometric.Cos where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

-- Cosine function
-- Mirrors Haskell Glue.Lib.Math.Trigonometric.Cos.cos exactly
cos :: IR Eval
cos = NativeFunc cosImpl

-- Cosine function implementation
-- Mirrors Haskell Glue.Lib.Math.Trigonometric.Cos.cosImpl exactly
cosImpl :: IR Eval -> Eval (IR Eval)
cosImpl arg = case arg of
    Integer n -> pure $ Float (Prelude.cos (fromIntegral n))
    Float n -> pure $ Float (Prelude.cos n)
    _ -> throwError $ wrongArgumentType ["number"]

File: dart/glue/lib/src/lib/math/trigonometric/cos.dart
import 'dart:math' as math;

import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Cosine function (radians)
/// Mirrors Haskell Glue.Lib.Math.Trigonometric.Cos.cos exactly
final Ir cos = IrNativeFunc(cosImpl);

/// Cosine function implementation (radians)
/// Mirrors Haskell Glue.Lib.Math.Trigonometric.Cos.cosImpl exactly
Eval<Ir> cosImpl(Ir arg) {
  return eval(arg).flatMap((va) {
    return switch (va) {
      IrInteger(value: final n) => Eval.pure(IrFloat(math.cos(n.toDouble()))),
      IrFloat(value: final n) => Eval.pure(IrFloat(math.cos(n))),
      _ => throwError(wrongArgumentType(['number'])),
    };
  });
}
